### ISSUE TITLE
suggestion: VirusTotal scan

### ISSUE THREAD

**Issue by DanielRuf** [phpMussel/phpMussel#12]
if the  uploaded file is not recognized by any ClamAV signature and a special constant in the settings of phpMussel is set, we could use the MD5 / Sha1 / SHA256 hash of the file and fetch some results from the VirusTotal API if there are any

https://www.virustotal.com/de/documentation/public-api/#getting-file-scans
http://stackoverflow.com/questions/5572043/php-hash-contents-of-uploaded-file

If this idea is a useful suggestion and addition, I would prepare this and test it.

**Comment by Maikuolan** [comment]
I definitely like the idea, and it's something that I'd thought about doing before but haven't yet tried doing at all or looked into too much. I can definitely see very significant benefits for phpMussel by including a feature like this though.

If we did implement something like this, of course, because using the Virus Total API requires registration of an API key, I think a feature like this would need to be disabled by default, having it as something that could be enabled by setting/defining an API key for it in the phpMussel configuration file.

We'd probably also need to document the feature very thoroughly in the documentation, to make sure that anyone using it understood exactly what it was and exactly how it worked, including some sort of clear warning that using the feature would involve communicating with their [the Virus Total] servers/services (because of the risk of users potentially panicking when they see phpMussel making outbound requests, panicking that phpMussel might be sharing their private documents with some server somewhere and things like that; they should already be aware of things like that if they register an API key and start using it, but, some people will always just go ahead, register something and start using it sometimes, without knowing what it's for or what it does), and probably also some message similar to "by using the Virus Total API feature of phpMussel, you hereby agree to the Virus Total <a href="https://www.virustotal.com/en/about/terms-of-service/">Terms of Service</a>", to keep Virus Total happy and to avoid anyone potentially trying to blame us if anything goes wrong.

I also spotted these on Virus Total's API documentation page; They might potentially be useful:

https://github.com/jayzeng/virustotal_apiwrapper
http://www.ab-weblog.com/en/php-script-for-accessing-virustotal-api-version-2-0/

I've briefly looked at them, but haven't pulled them apart yet to see what's in them, but, they might be worth looking at.

In their documentation, they also write this:

<em>"The public API is a free service, available for any website or application that is free to consumers. <b>The API must not be used in commercial products or services, it can not be used as a substitute for antivirus products and it can not be integrated in any project that may harm the antivirus industry directly or indirectly. Noncompliance of these terms will result in immediate permanent ban of the infractor individual or organization.</b> Please see the terms of service for more information."</em>

I'm reasonably certain that writing a feature into phpMussel to check files or hashes against Virus Total should okay, and shouldn't be in breech of their terms, but, I'm thinking it might be good idea, before starting any work on a feature like this, to send them a quick email explaining what we want to do and asking them if they'd be okay with it, just to be absolutely sure; Obviously, our intentions are good, and because phpMussel is open source GNU/GPLv2, we wouldn't be considered commercial, but, <em>"it can not be used as a substitute for antivirus products"</em> could be interpreted in a number of different ways, and depending on exactly what they intend for that to mean, they may or may not be okay with what we want to do.

Anyhow, if you're happy to follow up with this and happy to try putting something together, go for it; I'd love to see something like this implemented. :-)

**Comment by DanielRuf** [comment]
Right.

There are two ways to get infos about a file.
Either just get the report using a file hash (no file content is submitted) or upload the file to VirusTotal and then get the report.

I just thought about the first one as this is faster and might be something like a cloud reputation scanning method.

The second way is the slower and more exact way.

I was thinking the same, better ask the VirusTotal team before implementing this just to be sure.

**Comment by karlhiramoto** [comment]
+1  :smiley:

**Comment by DanielRuf** [comment]
@karlhiramoto awesome to have you here, so we have now some VirusTotal contact in case of further questions about the API and its implementation. Thanks =)

**Comment by Maikuolan** [comment]
Been busy trying to code quite a few new things into the core script over the past few days, including which is Virus Total integration; I'm still double-checking over everything to make there aren't any obvious errors before I commit, but, as of a couple of minutes ago, I think I may possibly have it working correctly. :D

(( @DanielRuf: I know we were going to wait a while to plan everything out, and I know that you were going to put something together for us, but, once I'd finished putting together the long overdue caching system last night, ways to go about the Virus Total integration began to emerge, and having quiet day today without much to do, I gave in to temptation; Sorry for stealing the task. I'd still be very keen to have you look over it once I've committed it to see if you can find ways to improve on it and things like that though, and, so far, although I've written code to check hashes against Virus Total, I haven't yet written code to send new files to Virus Total that they don't yet have reports for or code for anything else related to the integration, so, there's all that, still. :P ))

![virustotalscreenshot](https://cloud.githubusercontent.com/assets/12571108/8276655/99b10556-18eb-11e5-8faf-533303f4b8e6.png)

**Comment by DanielRuf** [comment]
@Maikuolan Awesome, I am very excited to see the upcoming VirusTotal implementation. =)

No problem, it is a great pleasure to help with ideas, the development and contribute to this project in different ways.

Sure, I will look over the code when it is committed and make notes about further improvements.

**Comment by Maikuolan** [comment]
@DanielRuf Excellent. :-)

Commit has been sent. I haven't had time to deal properly with translating all the new documentation content yet, but I'll get that sorted out at some point during the week (maybe you could take a look at the new German documentation content? That would definitely help a lot. I've dumped an English version of the new content into all the files for the time being).

@karlhiramoto Woohoo, it's working (at least, so far as I can tell; I'll leave it marked as BETA at least until there's been a bit more testing)! ^_^

If interested and if you've got the spare time, I'd love to hear what you think (no rush, of course). :-)

**Comment by DanielRuf** [comment]
@Maikuolan I think you have sometimes written Virus Total and sometimes VirusTotal, the second one is right, I think. Nothing serious and this could be changed very fast. Just wanted to mention this.

I have checked the VirusTotal implementation and so far I am satisfied with it and I really like it. Unfortunately the time for more testing is not available at the moment but I will do it when I have the time for it (using my collected malware samples and stress test phpMussel with them, some big malware repositories are also available on the internet which is good for testing).

**Comment by Maikuolan** [comment]
Good catch. I'll get onto that early tomorrow when I continue with the documentation again.

No rush, of course, but, looking forward to seeing the results of the stress test. :-)

**Comment by DanielRuf** [comment]
@karlhiramoto What is the size of the biggest (uploaded) and analysed file on VirusTotal?

The website says the maximum filesize for the upload is 128MB.

@Maikuolan Do we already check this restriction?

**Comment by Maikuolan** [comment]
We don't specifically, so, that's definitely something that needs to be added in (won't be a problem, per se, until the ability to send/upload actual files to Virus Total is coded for, which hasn't happened yet, but, it is on the to-do list, so, checking the filesize restrictions for uploads sent to Virus Total is something that should also be done for when the time comes); That said, though, it wouldn't be possible to scan a file larger than whatever restrictions phpMussel itself placed on files to be scanned (phpMussel's `filesize_limit` directive defaults to 64MB, and the default response for files that exceed the `filesize_limit` directive is to consider them blacklisted and block them from being uploaded), due to the need to read the file contents in order to calculate the MD5 to send in a request to Virus Total. Whether or not the default response to those files exceeding the `filesize_limit` directive would have an effect on the filesize restriction that Virus Total has for uploaded file, though, I guess, would depend on how we choose to eventually implement the ability to send/upload files to Virus Total (whether this is done before or after phpMussel detects is as malicious/benign, what sort of checks are in effect, whether or not we send the files after a decision has been made by phpMussel, whether or not we send them when phpMussel can't make a decision about a file, whether or not we restrict sending new files to Virus Total to those phpMussel installations that enable the `honeypot_mode` directive, and then, if we do, whether or not we're sending -all- files, or just those that are considered "suspicious"; I guess there are quite a few things to decide, but, we'll work it out, lol).

**Comment by DanielRuf** [comment]
I was a bit unprecise, sorry. 

I meant that we do not have to check the hash with the VirusTotal API if the filesize exceeds the current limit of 128 MB per file (a single file eg in an archive).

I think the public API has a limit of 64 MB per file for upload.

But yes, we need to check this even for the planned implementation for uploading files to VirusTotal.

The cache works great by the way.

**Comment by Maikuolan** [comment]
Ah, okay. Good point.

**Comment by karlhiramoto** [comment]
the max size of file/scan API call is 32mb. there is a different API call for bigger files

**Comment by DanielRuf** [comment]
> the max size of file/scan API call is 32mb. there is a different API call for bigger files

Ah ok, thanks for the clarification.

**Comment by Maikuolan** [comment]
_Removing Testing label; Current implementation of the Virus Total integration seems thoroughly stable at this point. Keeping this issue open due to having not yet implemented a way for phpMussel to send new binaries back to Virus Total to queue for later scanning; This should be a considered a to-do item._

**Comment by Maikuolan** [comment]
Will create a new issue to deal with sending samples back to Virus Total, so that we can close this old issue.
